runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java

```diff
@@ -299,18 +304,33 @@ public void invalidateClientCache() {
      * @param addresses collection of addresses to read from.
      * @return A result to be cached
      */
-    public @Nonnull Map<Long, ILogData> cacheFetch(Iterable<Long> addresses) {
-        //turn the addresses into Set for now to satisfy signature requirement down the line
-        Set<Long> readAddresses = new TreeSet<>();
-        Iterator<Long> iterator = addresses.iterator();
-        while(iterator.hasNext()){
-            readAddresses.add(iterator.next());
+    public @Nonnull
+    Map<Long, ILogData> cacheFetch(Iterable<Long> addresses) {
+        Map<Long, ILogData> temp = new HashMap<>();
+
+        Iterable<List<Long>> batches = Iterables.partition(addresses, runtime.getBulkReadSize());
+
+        for (List<Long> batch : batches) {
+            //doesn't handle the case where some address have a different replication mode
+            temp.putAll(layoutHelper(l -> l.getReplicationMode(batch.iterator().next())
+                    .getReplicationProtocol(runtime)
+                    .readAll(l, batch)));
```

Comment:
what happens when the read for a batch fails ? Should we fail the whole large batch. Also is there a need to make this serial ? The result itself is a Map which has no requirement like that.